File: src/main/java/com/joecollins/graphics/GenericPanelWithHeaderAndLowerThird.java
package com.joecollins.graphics;

import com.joecollins.bindings.Binding;
import com.joecollins.graphics.components.lowerthird.LowerThird;
import com.joecollins.graphics.utils.StandardFont;
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Dimension;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.border.EmptyBorder;

public class GenericPanelWithHeaderAndLowerThird<T extends JPanel> extends JPanel {

  private final T panel;
  private final JLabel label;
  private final LowerThird lowerThird;

  public GenericPanelWithHeaderAndLowerThird(T panel, String label) {
    this(panel, Binding.fixedBinding(label), null);
  }

  public GenericPanelWithHeaderAndLowerThird(T panel, LowerThird lowerThird) {
    this(panel, (Binding<String>) null, lowerThird);
  }

  public GenericPanelWithHeaderAndLowerThird(T panel, String label, LowerThird lowerThird) {
    this(panel, Binding.fixedBinding(label), lowerThird);
  }

  public GenericPanelWithHeaderAndLowerThird(
      T panel, Binding<String> label, LowerThird lowerThird) {
    this.panel = panel;
    this.lowerThird = lowerThird;
    this.label = new JLabel();
    if (label != null) {
      label.bind(this.label::setText);
      this.label.setHorizontalAlignment(JLabel.CENTER);
      this.label.setBorder(new EmptyBorder(5, 0, -5, 0));
      this.label.setFont(StandardFont.readBoldFont(32));
    }

    setBackground(Color.WHITE);
    setLayout(new BorderLayout());
    setPreferredSize(new Dimension(1024, 512));
    if (label != null) {
      add(this.label, BorderLayout.NORTH);
    }
    if (lowerThird != null) {
      add(this.lowerThird, BorderLayout.SOUTH);
    }
    add(this.panel, BorderLayout.CENTER);
  }

  public T getPanel() {
    return panel;
  }

  public JLabel getLabel() {
    return label;
  }

  public LowerThird getLowerThird() {
    return lowerThird;
  }
}

File: src/main/java/com/joecollins/graphics/GenericPanelWithHeaderAndLowerThird.kt
package com.joecollins.graphics

import com.joecollins.graphics.components.lowerthird.LowerThird
import com.joecollins.graphics.utils.StandardFont.readBoldFont
import com.joecollins.pubsub.Subscriber
import com.joecollins.pubsub.Subscriber.Companion.eventQueueWrapper
import com.joecollins.pubsub.asOneTimePublisher
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Dimension
import java.util.concurrent.Flow
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.border.EmptyBorder

class GenericPanelWithHeaderAndLowerThird<T : JPanel>(
    panel: T,
    label: Flow.Publisher<String>?,
    lowerThird: LowerThird?
) : JPanel() {

    private val _label: JLabel = JLabel()

    constructor(panel: T, label: String) : this(panel, label.asOneTimePublisher(), null)
    constructor(panel: T, lowerThird: LowerThird?) : this(panel, null as Flow.Publisher<String>?, lowerThird)
    constructor(panel: T, label: String, lowerThird: LowerThird?) : this(panel, label.asOneTimePublisher(), lowerThird)

    init {
        if (label != null) {
            label.subscribe(Subscriber(eventQueueWrapper { text: String? -> this._label.text = text ?: "" }))
            this._label.horizontalAlignment = JLabel.CENTER
            this._label.border = EmptyBorder(5, 0, -5, 0)
            this._label.font = readBoldFont(32)
        }
        background = Color.WHITE
        layout = BorderLayout()
        preferredSize = Dimension(1024, 512)
        if (label != null) {
            add(this._label, BorderLayout.NORTH)
        }
        if (lowerThird != null) {
            add(lowerThird, BorderLayout.SOUTH)
        }
        add(panel, BorderLayout.CENTER)
    }
}
